### PR TITLE
feat: Support calling influxql/flux selector aggregates from IOx SQL

### PR DESCRIPTION
**Pull request by alamb**
Closes https://github.com/influxdata/influxdb_iox/issues/5528

Note the vast majority of this PR is tests. The rest of basically plumbing.

# Rationale:
The flux team needs (https://github.com/influxdata/idpe/issues/15748 https://github.com/influxdata/idpe/pull/15664) this to be able to push aggregates down in  flux sql

# Changes
- [x] Refactor selectors to provide their output as `structs` in addition to the individual fields / values
- [x] Register `first`, `last`, `min` and `max`  as two function (e.g. `first_value` and `first_time`)
- [x] SQL based tests

# TODO
- [x] Merge https://github.com/apache/arrow-datafusion/pull/3428
- [x] Update datafusion pin to include https://github.com/apache/arrow-datafusion/pull/3428
- [x] Finish testing for other types
- [x] Port over other selectors to new format
- [x] Remove old selector implementations
- [x] Remove code duplication
- [x] File upstream DataFusion ticket for better API to register user defined functions: https://github.com/apache/arrow-datafusion/issues/3465

### CODE REVIEW

File: query_functions/src/selectors.rs [hunk 1]
```diff
@@ -147,15 +251,129 @@ pub fn selector_max(data_type: &DataType, output: SelectorOutput) -> AggregateUD
...
-        DataType::Utf8 => make_uda::<Utf8MaxSelector>(name, output),
-        DataType::Boolean => make_uda::<BooleanMaxSelector>(name, output),
-        _ => unimplemented!("max not supported for {:?}", data_type),
+    make_uda(
+        name,
+        FactoryBuilder::new(SelectorType::Max, output).with_value_type(data_type.clone()),
+    )
+}
+
+#[derive(Debug, Clone, Copy)]
```
Comment (alamb) [inline query_functions/src/selectors.rs:263]: This was largely a plumbing exercise to move instantiation of the selectors into code that didn't have too much duplication

File: query_functions/src/selectors.rs [hunk 2]
```diff
@@ -35,6 +41,113 @@ use internal::{
...
 use schema::TIME_DATA_TYPE;
 
+/// registers selector functions so they can be invoked via SQL
+pub fn register_selector_aggregates(mut state: SessionState) -> SessionState {
+    let first = struct_selector_first();
+    let last = struct_selector_last();
+    let min = struct_selector_min();
+    let max = struct_selector_max();
+
+    //TODO make a nicer api for this in DataFusion
```
Comment (alamb) [inline query_functions/src/selectors.rs:51]:
```suggestion
    // TODO make a nicer api for this in DataFusion
    // https://github.com/apache/arrow-datafusion/issues/3465
```

File: query_tests/cases/in/selectors.sql [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+select selector_first(float_field, time) from m;
+select selector_first(int_field, time) from m;
+select selector_first(uint_field, time) from m;
+select selector_first(string_field, time) from m;
+select selector_first(bool_field, time) from m;
+
+-- Test extracting value and time subfields
+select selector_first(float_field, time)['value'], selector_first(float_field, time)['time'] from m;
+-- Also with subqueries
+select f['value'], f['time'] from (select selector_first(float_field, time) as f from m) as sq;
```
Comment (alamb) [inline query_tests/cases/in/selectors.sql:21]: here are the SQL tests showing how to use these selector functions and their struct output

File: query_functions/src/selectors.rs [hunk 4]
```diff
@@ -54,16 +167,13 @@ pub fn selector_first(data_type: &DataType, output: SelectorOutput) -> Aggregate
...
-        DataType::Int64 => make_uda::<I64FirstSelector>(name, output),
-        DataType::UInt64 => make_uda::<U64FirstSelector>(name, output),
-        DataType::Utf8 => make_uda::<Utf8FirstSelector>(name, output),
-        DataType::Boolean => make_uda::<BooleanFirstSelector>(name, output),
-        _ => unimplemented!("first not supported for {:?}", data_type),
-    }
+    make_uda(
+        name,
+        FactoryBuilder::new(SelectorType::First, output).with_value_type(data_type.clone()),
+    )
```
Comment (NGA-TRAN) [inline query_functions/src/selectors.rs:176]: 🎉

File: query_functions/src/selectors.rs [hunk 5]
```diff
@@ -147,15 +251,129 @@ pub fn selector_max(data_type: &DataType, output: SelectorOutput) -> AggregateUD
...
+
+/// Builder to create the appropriate typed factory functions for selectors
+#[derive(Debug)]
+struct FactoryBuilder {
+    selector_type: SelectorType,
+
+    output_type: SelectorOutput,
+
+    // If the selector output is "time" we can't determine the
+    // accumuator type from the return type, so hold we pass the data type explicitly
```
Comment (NGA-TRAN) [inline query_functions/src/selectors.rs:279]: question: can the return type be the same`time` or it must be different?

File: query_tests/cases/in/selectors.expected [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
++--------------------------------------------------------+
+| {"value": true, "time": 1970-01-01 00:00:00.000000600} |
++--------------------------------------------------------+
+-- SQL: select selector_last(float_field, time)['value'], selector_last(float_field, time)['time'] from m;
++--------------------------------------------+-------------------------------------------+
+| selector_last(m.float_field,m.time)[value] | selector_last(m.float_field,m.time)[time] |
++--------------------------------------------+-------------------------------------------+
+| 66                                         | 1970-01-01T00:00:00.000000600Z            |
++--------------------------------------------+-------------------------------------------+
+-- SQL: select f['value'], f['time'] from (select selector_last(float_field, time) as f from m) as sq;
```
Comment (NGA-TRAN) [inline query_tests/cases/in/selectors.expected:93]: This query is very interesting

**Review by NGA-TRAN — APPROVED** [review]
Very nice. Now I know how to use the selector function. I went over all the functions and tests carefully.

**Comment by alamb** [comment]
Thank you for the careful review @NGA-TRAN 👍

File: query_tests/cases/in/selectors.expected (hunk 6 above)
Comment (alamb) [reply query_tests/cases/in/selectors.expected:93]: I was (very) pleasantly surprised to see that the `['field']`syntax worked and that enough of DataFusion supported these struct types to use.

File: query_functions/src/selectors.rs (hunk 5 above)
Comment (alamb) [reply query_functions/src/selectors.rs:279]: The return type can be the same as time (aka `DataType::Timestamp(TimeUnit::Nanoseconds, None)`) . However the accumulator state needs to hold both the `value` and the `time`

The factory function that figures out the accumulator's state type is passed the return type. When the output contains the actual column value (e.g. the `i64`/ `u64`, etc) then the internal state can be recovered from the return type. However, when the output type is just `time` then there needs to be some other way to know the column value's type

I actually tried to change the selectors to *always* return a struct (that is what https://github.com/influxdata/influxdb_iox/pull/5606 does) but I ran into trouble making the influxrpc interface use them.

**Pull request merged**